File: Documentation/5.4/Samples/python/Indexes/Querying/Intersection.py
from typing import List

from ravendb import AbstractIndexCreationTask

from examples_base import ExampleBase


# region intersection_1
class TShirtType:
    def __init__(self, color: str = None, size: str = None):
        self.color = color
        self.size = size


class TShirt:
    def __init__(
        self, Id: str = None, release_year: int = None, manufacturer: str = None, types: List[TShirtType] = None
    ):
        self.Id = Id
        self.release_year = release_year
        self.manufacturer = manufacturer
        self.types = types


# endregion


# region intersection_2
class TShirts_ByManufacturerColorSizeAndReleaseYear(AbstractIndexCreationTask):
    class Result:
        def __init__(self, manufacturer: str = None, color: str = None, size: str = None, release_year: int = None):
            self.manufacturer = manufacturer
            self.color = color
            self.size = size
            self.release_year = release_year

    def __init__(self):
        super().__init__()
        self.map = (
            "from tshirt in docs.TShirts from type in tshirt.types select new {"
            "  manufacturer = tshirt.manufacturer,"
            "  color = tshirt.color,"
            "  size = type.size,"
            "  release_year = tshirt.release_year"
            "}"
        )


# endregion


class Intersection(ExampleBase):
    def setUp(self):
        super().setUp()

    def test_sample(self):
        with self.embedded_server.get_document_store("Intersection") as store:
            with store.open_session() as session:
                # region intersection_3
                session.store(
                    TShirt(
                        Id="tshirts/1",
                        manufacturer="Raven",
                        release_year=2010,
                        types=[
                            TShirtType(color="Blue", size="Small"),
                            TShirtType(color="Black", size="Small"),
                            TShirtType(color="Black", size="Medium"),
                            TShirtType(color="Gray", size="Large"),
                        ],
                    )
                )

                session.store(
                    TShirt(
                        Id="tshirts/2",
                        manufacturer="Wolf",
                        release_year=2011,
                        types=[
                            TShirtType(color="Blue", size="Small"),
                            TShirtType(color="Black", size="Large"),
                            TShirtType(color="Gray", size="Medium"),
                        ],
                    )
                )

                session.store(
                    TShirt(
                        Id="tshirts/3",
                        manufacturer="Raven",
                        release_year=2011,
                        types=[TShirtType(color="Yellow", size="Small"), TShirtType(color="Gray", size="Large")],
                    )
                )

                session.store(
                    TShirt(
                        Id="tshirts/4",
                        manufacturer="Raven",
                        release_year=2012,
                        types=[TShirtType(color="Blue", size="Small"), TShirtType(color="Gray", size="Large")],
                    )
                )
                # endregion
                session.save_changes()
            with store.open_session() as session:
                # region intersection_4
                results = list(
                    session.query_index_type(
                        TShirts_ByManufacturerColorSizeAndReleaseYear,
                        TShirts_ByManufacturerColorSizeAndReleaseYear.Result,
                    )
                    .where_equals("Manufacturer", "Raven")
                    .intersect()
                    .where_equals("Color", "Blue")
                    .and_also()
                    .where_equals("Size", "Small")
                    .intersect()
                    .where_equals("Color", "Gray")
                    .and_also()
                    .where_equals("Size", "Large")
                )

                # endregion

File: Documentation/5.4/Samples/php/Indexes/Querying/Intersection.php
<?php

namespace RavenDB\Samples\Indexes\Querying;

use RavenDB\Documents\DocumentStore;
use RavenDB\Documents\Indexes\AbstractIndexCreationTask;
use RavenDB\Type\TypedList;

class Intersection
{
    public function sample(): void
    {
        $store = new DocumentStore();
        try {
            $session = $store->openSession();
            try {
                # region intersection_3

                // first tShirt
                $tShirt = new TShirt();
                $tShirt->setId("tshirts/1");
                $tShirt->setManufacturer("Raven");
                $tShirt->setReleaseYear(2010);

                $types = TShirtTypeList::fromArray([
                    new TShirtType(color: "Blue", size: "Small"),
                    new TShirtType(color: "Black", size: "Small"),
                    new TShirtType(color: "Black", size: "Medium"),
                    new TShirtType(color: "Gray", size: "Large")
                ]);
                $tShirt->setTypes($types);

                $session->store($tShirt);

                // second tShirt
                $tShirt = new TShirt();
                $tShirt->setId("tshirts/2");
                $tShirt->setManufacturer("Wolf");
                $tShirt->setReleaseYear(2011);

                $types = TShirtTypeList::fromArray([
                    new TShirtType(color: "Blue", size: "Small"),
                    new TShirtType(color: "Black", size: "Large"),
                    new TShirtType(color: "Black", size: "Medium")
                ]);
                $tShirt->setTypes($types);

                $session->store($tShirt);

                // third tShirt
                $tShirt = new TShirt();
                $tShirt->setId("tshirts/3");
                $tShirt->setManufacturer("Raven");
                $tShirt->setReleaseYear(2011);

                $types = TShirtTypeList::fromArray([
                    new TShirtType(color: "Yellow", size: "Small"),
                    new TShirtType(color: "Gray", size: "Large")
                ]);
                $tShirt->setTypes($types);

                $session->store($tShirt);

                // fourth tShirt
                $tShirt = new TShirt();
                $tShirt->setId("tshirts/4");
                $tShirt->setManufacturer("Raven");
                $tShirt->setReleaseYear(2012);

                $types = TShirtTypeList::fromArray([
                    new TShirtType(color: "Blue", size: "Small"),
                    new TShirtType(color: "Gray", size: "Large")
                ]);
                $tShirt->setTypes($types);

                $session->store($tShirt);
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region intersection_4
                /** @var array<TShirt> $results */
                $results = $session->query(TShirts_ByManufacturerColorSizeAndReleaseYear_Result::class, TShirts_ByManufacturerColorSizeAndReleaseYear::class)
                    ->whereEquals("Manufacturer", "Raven")
                    ->intersect()
                    ->whereEquals("Color", "Blue")
                    ->andAlso()
                    ->whereEquals("Size", "Small")
                    ->intersect()
                    ->whereEquals("Color", "Gray")
                    ->andAlso()
                    ->whereEquals("Size", "Large")
                    ->ofType(TShirt::class)
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }

            $session = $store->openSession();
            try {
                # region intersection_5
                /** @var array<TShirt> $results */
                $results = $session
                    ->advanced()
                    ->documentQuery(TShirt::class, TShirts_ByManufacturerColorSizeAndReleaseYear::class)
                    ->whereEquals("Manufacturer", "Raven")
                    ->intersect()
                    ->whereEquals("Color", "Blue")
                    ->andAlso()
                    ->whereEquals("Size", "Small")
                    ->intersect()
                    ->whereEquals("Color", "Gray")
                    ->andAlso()
                    ->whereEquals("Size", "Large")
                    ->toList();
                # endregion
            } finally {
                $session->close();
            }
        } finally {
            $store->close();
        }
    }
}

# region intersection_1
class TShirt
{
    private ?string $id = null;
    private ?int $releaseYear = null;
    private ?string $manufacturer = null;
    private ?TShirtTypeList $types = null;

    public function getId(): ?string
    {
        return $this->id;
    }

    public function setId(?string $id): void
    {
        $this->id = $id;
    }

    public function getReleaseYear(): ?int
    {
        return $this->releaseYear;
    }

    public function setReleaseYear(?int $releaseYear): void
    {
        $this->releaseYear = $releaseYear;
    }

    public function getManufacturer(): ?string
    {
        return $this->manufacturer;
    }

    public function setManufacturer(?string $manufacturer): void
    {
        $this->manufacturer = $manufacturer;
    }

    public function getTypes(): ?TShirtTypeList
    {
        return $this->types;
    }

    public function setTypes(?TShirtTypeList $types): void
    {
        $this->types = $types;
    }
}

class TShirtType
{
    private ?string $color = null;
    private ?string $size = null;

    public function __construct(?string $color, ?string $size)
    {
        $this->color = $color;
        $this->size = $size;
    }

    public function getColor(): ?string
    {
        return $this->color;
    }

    public function setColor(?string $color): void
    {
        $this->color = $color;
    }

    public function getSize(): ?string
    {
        return $this->size;
    }

    public function setSize(?string $size): void
    {
        $this->size = $size;
    }
}

class TShirtTypeList extends TypedList
{
    protected function __construct()
    {
        parent::__construct(TShirtType::class);
    }
}
# endregion

# region intersection_2
class TShirts_ByManufacturerColorSizeAndReleaseYear_Result
{
    private ?string $manufacturer = null;
    private ?string $color = null;
    private ?string $size = null;
    private ?int $releaseYear = null;
    // ... getters and setters
}
class TShirts_ByManufacturerColorSizeAndReleaseYear extends AbstractIndexCreationTask
{
    public function __construct()
    {
        parent::__construct();

        $this->map = "docs.TShirts.SelectMany(tshirt => tshirt.types, (tshirt, type) => new {" .
            "    manufacturer = tshirt.manufacturer," .
            "    color = type.color," .
            "    size = type.size," .
            "    releaseYear = tshirt.releaseYear" .
            "})";
    }
}
# endregion
